spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/internal/spring/ArmeriaConfigurationUtil.java

```diff
@@ -201,15 +182,62 @@ public static void configureServerWithArmeriaSettings(
         }
     }
 
-    private static boolean hasAllClasses(String... classNames) {
-        for (String className : classNames) {
-            try {
-                Class.forName(className, false, ArmeriaConfigurationUtil.class.getClassLoader());
-            } catch (Throwable t) {
-                return false;
+    private static void configureInternalService(ServerBuilder server, InternalServiceId serviceId,
+                                                 @Nullable String servicePath,
+                                                 @Nullable HttpService service,
+                                                 List<Port> internalPorts,
+                                                 @Nullable List<InternalServiceId> internalServiceIds,
+                                                 boolean usesPrefixPath) {
+        if (service == null) {
+            return;
+        }
+        // An internal service should be created only when a servicePath is not null.
+        assert servicePath != null;
+        internalServiceIds = firstNonNull(internalServiceIds, ImmutableList.of());
+
+        if (internalPorts.isEmpty()) {
+            // No internal ports are configured. The default virtual is used to use the service.
+            configureInternalService(server, serviceId, servicePath, service,
+                                     (Port) null, internalServiceIds, usesPrefixPath);
+        } else {
+            for (Port internalPort : internalPorts) {
+                configureInternalService(server, serviceId, servicePath, service,
+                                         internalPort, internalServiceIds, usesPrefixPath);
+            }
+        }
+    }
+
+    private static void configureInternalService(ServerBuilder server, InternalServiceId serviceId,
+                                                 String servicePath,
+                                                 HttpService service, @Nullable Port internalPort,
+                                                 List<InternalServiceId> internalServiceIds,
+                                                 boolean usesPrefixPath) {
+        final boolean needsPortBasedVirtualHost;
+        if (internalPort == null) {
+            needsPortBasedVirtualHost = false;
+        } else {
+            if (internalServiceIds.size() == 1 && internalServiceIds.get(0) == InternalServiceId.ALL) {
```

Comment:
How about just:
```suggestion
            if (internalServiceIds.contains(InternalServiceId.ALL)) {
```
in case ALL and others are specified together?

Reply:
That could be a misconfiguration. Should it be better to raise an exception?

Reply:
On the second thought, make sense to use `contains`